### ISSUE TITLE
t.w.xish.domish.Element recursion error

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2880]
|<img alt="twonds's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| twonds reported|
|-|-|
|Trac ID|trac#2880|
|Type|defect|
|Created|2007-11-09 17:29:41Z|

When you pickle a domish.Element object you get the following error. 'RuntimeError: maximum recursion depth exceeded' 

A fix is attached, unit tests to follow. 

Attachments:

* [tw.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/874/8742c668ee52f7cbe5181d609ff800f3a37492c5/a199ee5e1af11f74c5e0ff2ac5cb371e7759da5a.diff) (637 bytes) - added by twonds on 2007-11-09 17:29:58Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__2880 2880
type__defect defect
reporter__twonds twonds
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__new new
resolution__None None
component__words words
keywords__ 
time__1194629381000000 1194629381000000
changetime__1265387458000000 1265387458000000
version__None None
owner__ 
cc__jack cc__ralphm cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

This behavior of Element seems highly dubious to me.  My inclination at this point is to deprecate !__getattr!__.

Of course, I really, really don't understand why xish exists at all.  Maybe there's a perfectly reasonable explanation for this particular API (although I doubt it) which someone could explain.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(Deprecate, in addition to making it work in this case, that is.)

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_3"></a>|@ralphm commented|
|-|-|

First off, `Element` was never designed to be pickled. As mentioned on the mailing list, the obvious way to store (partial) DOMs in persistent storage is by retrieving their serialization as an XML document.

On why xish exists: it was built solely for streaming XML, with Jabber/XMPP as its primary target. I have argued on several occasions on some of the design decisions and why to keep things as they are in xish. I'll try to list it shortly below:

* XMPP has certain constraints on serializations, documented in [RFC 3920, section 11](http://www.xmpp.org/rfcs/rfc3920.html#xml).
Other XML library implementations implemented in Python or with Python bindings (e.g. ElementTree, libxml2, lxml) currently do not provide sufficient control over serialization to meet these constraints. The serialization code in `domish` does, and has a large number of test cases to back it up. The DOM representation in the form of `Element` provides an API that is efficient for acting on the XML that is typical for XMPP exchanges.
* Streaming XML works by passing incoming traffic to an XML parser (`domish` supports both Expat and `twisted.web.sux`) and acting on partial DOMs that are created from the events produced by the parser. The unit of communication is first level elements of the root element dubbed XML Stanzas. `t.w.x.utility.EventDispatcher` is provided to call callbacks on such XML Stanzas based on XPath-like queries.
* The XPath-like implementation is far from complete and does not fully comply with the XPath specification, but is more than sufficient for the typical XMPP use cases that it was created for.

There have been investigations into and experiments with refitting the core of xish with an existing XML library to reduce the amount of code in xish, have one less API and maybe increase processing speed. E.g. jack (cc'd) of Chesspark worked on this for a while. It turns out that serialization is comparable in speed with C based serializations and the XPath code is unmatched in speed. The later is most likely because of the efficiency of the yapps scanner/parser and the incompleteness of the implementation.

In conclusion, so far it seems that xish meets the needs of its users and I have yet to find compelling reasons to do away with it. But I remain open for discussion.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

When I said _for this particular API_, I meant ```Element.__getattr__```.  Sorry for being unclear, but thanks for the overview.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

I think this is a duplicate of [#7246](https://github.com/twisted/twisted/issues/7246).  Can anyone confirm?

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
